src/Compilers/CSharp/Test/Semantic/SourceGeneration/TestGenerators.cs

```diff
@@ -0,0 +1,107 @@
+    {
+        private readonly string _content;
+        private readonly string _hintName;
+
+        public SingleFileTestGenerator(string content, string hintName = "generatedFile")
+        {
+            _content = content;
+            _hintName = hintName;
+        }
+
+        public void Execute(SourceGeneratorContext context)
+        {
+            context.AdditionalSources.Add(this._hintName, SourceText.From(_content, Encoding.UTF8));
+        }
+
+        public void Initialize(InitializationContext context)
+        {
+        }
+    }
+
+    internal class CallbackGenerator : ISourceGenerator
+    {
+        private readonly Action<InitializationContext> _onInit;
+        private readonly Action<SourceGeneratorContext> _onExecute;
+        private readonly string _sourceOpt;
+
+        public CallbackGenerator(Action<InitializationContext> onInit, Action<SourceGeneratorContext> onExecute, string sourceOpt = "")
+        {
+            _onInit = onInit;
+            _onExecute = onExecute;
+            _sourceOpt = sourceOpt;
+        }
+
+        public void Execute(SourceGeneratorContext context)
+        {
+            _onExecute(context);
+            if (!string.IsNullOrWhiteSpace(_sourceOpt))
+            {
+                context.AdditionalSources.Add("source.cs", SourceText.From(_sourceOpt, Encoding.UTF8));
+            }
+        }
+        public void Initialize(InitializationContext context) => _onInit(context);
+    }
+
+    internal class AdditionalFileAddedGenerator : ISourceGenerator
+    {
+        public bool CanApplyChanges { get; set; } = true;
+
+        public void Execute(SourceGeneratorContext context)
+        {
+            foreach (var file in context.AdditionalFiles)
+            {
+                AddSourceForAdditionalFile(context.AdditionalSources, file);
+            }
+        }
+
+        public void Initialize(InitializationContext context)
+        {
+            context.RegisterForAdditionalFileChanges(UpdateContext);
+        }
+
+        bool UpdateContext(EditContext context, AdditionalFileEdit edit)
+        {
+            if (edit is AdditionalFileAddedEdit add && CanApplyChanges)
+            {
+                AddSourceForAdditionalFile(context.AdditionalSources, add.AddedText);
+                return true;
+            }
+            return false;
+        }
+
+        private void AddSourceForAdditionalFile(AdditionalSourcesCollection sources, AdditionalText file) => sources.Add(GetGeneratedFileName(file.Path), SourceText.From("", Encoding.UTF8));
+
+        private string GetGeneratedFileName(string path) => $"{Path.GetFileNameWithoutExtension(path)}.generated";
+    }
+
+    internal class InMemoryAdditionalText : AdditionalText
+    {
+        private readonly SourceText _content;
+
+        public InMemoryAdditionalText(string path, string content)
+        {
+            Path = path;
+            _content = SourceText.From(content);
```

Comment:
In all other `SourceText.From` you explicitly passed `Encoding.Utf8`. Why is this usage different?

Reply:
Oversight. I'll fix it, but it only actually matters when we're emitting, so we don't hit this in the tests. We have an open tracking note on the board about how to simplify this.

---
In reply to: [434195219](https://github.com/dotnet/roslyn/pull/44768#discussion_r434195219) [](ancestors = 434195219)